pyquil/device/transformers/rpcq.py

```diff
@@ -0,0 +1,115 @@
+import numpy as np
+from typing import List, Union, Optional
+from qcs_api_client.models import InstructionSetArchitecture, Operation, OperationSite
+from pyquil.device.transformers._rpcq_types import (
+    TargetDevice, Qubit, Edge, GateInfo, MeasureInfo,
+)
+
+
+def add_qubit(device: TargetDevice, node_id: int):
+    if node_id not in device.qubits:
+        device.qubits[str(node_id)] = Qubit()
+
+
+def get_qubit(device: TargetDevice, node_id: int) -> Optional[Qubit]:
+    return device.qubits.get(str(node_id))
+
+
+def make_edge_id(qubit1: int, qubit2: int) -> str:
+    return '-'.join([str(qubit) for qubit in sorted([qubit1, qubit2])])
+
+
+def add_edge(device: TargetDevice, qubit1: int, qubit2: int):
+    edge_id = make_edge_id(qubit1, qubit2)
+    if edge_id not in device.edges:
+        device.edges[edge_id] = Edge()
+
+
+def get_edge(device: TargetDevice, qubit1: int, qubit2: int) -> Optional[Qubit]:
+    return device.edges.get(make_edge_id(qubit1, qubit2))
+
+
+def transform_isa_to_target_device(isa: InstructionSetArchitecture) -> TargetDevice:
+    device = TargetDevice()
+    # QUESTION: Should we include qubits and edges that have no operations?
+    for node in isa.architecture.nodes:
+        add_qubit(device, node.node_id)
+
+    for edge in isa.architecture.edges:
+        add_edge(edge.node_ids[0], edge.node_ids[2])
+
+    for operation in isa.instructions:
+        for site in operation.sites:
+            if operation.node_count == 1:
+                # QUESTION: This used to return a default set of gates for a qubit.
+                qubit = get_qubit(device, site.node_ids[0])
+                # QUESTION: Need to check against redundant gates here?
+                qubit.gates.extend(transform_qubit_operation_to_gates(operation))
+
+            elif operation.node_count == 2:
+                edge = get_edge(device, site.node_ids[0], site.node_ids[1])
+                edge.gates.extend(transform_edge_operation_to_gates(operation))
+
+            else:
+                # QUESTION: Log error here? Include parameter for hard or soft failure?
+                raise ValueError("unexpected operation node count: {}".format(operation.node_count))
+    return device
+
+
+def transform_qubit_operation_to_gates(
+    operation: Operation, site: OperationSite,
+) -> List[Union[GateInfo, MeasureInfo]]:
+
+    # QUESTION: should we maintain a mapping here of operation to expected fidelity and
```

Comment:
To date, for my purposes, I've used an explicit mapping. In most ways I'm comfortable with that because these strings are proxies for very specific (if not easily visible to the user) methodologies for measuring them. @caldwellshane would be another good opinion here as I recall discussions with him when helping engineer the ISA around (a) different methods that could be used and reported to calculate fidelity.

While this is limited to a public discussion, I think its likely this list will grow over time and include different metrics, not just fidelities, and its perhaps fragile to assume that fidelity is the first one. I'd be pro- an explicit mapping for now, noting that it's one more thing to synchronize on rollout of gate set changes or additions (which is a drawback).

The other "good" approach is to standardize "fidelity" as the name for which at least this pyquil sense of fidelity needs to be mapped.

Reply:
For now I'm going to use a mapping in file, but I concur the ISAs should be fully self-described; that'll help reduce Pyquil updates.